### PR TITLE
Improve accessibility of the registration form

### LINKED ISSUE #6059: [A11Y] Registration form fields are not associated with the respective labels
**Describe the bug**

The form fields on the registration page are not associated with the respective labels. This includes (but not limited to):

**To Reproduce**
1. Go to a registration form using a screen reader
2. Focus any control
3. Observe that the field name is not announced

**Expected behavior**

The field name is announced.

**Additional context**

https://www.w3.org/WAI/WCAG21/Understanding/labels-or-instructions.html

### LINKED ISSUE #6073: [A11Y] Semantic UI date input overrides the associated label with placeholder text
**Describe the bug**

When using the date picker with a screen reader, the input is always read as if the placeholder is the label, regardless of whether there is a proper field label associated with it.

**To Reproduce**
1. Go to any form that includes a date picker
2. If necessary, manually associate the label with the input using a developer tool (usually they aren't associated)
3. While using a screen reader, focus the date input
4. Observe that the field is announced as "why why why why em em dee dee"

**Expected behavior**

The associated label is read

**Additional context**
https://www.w3.org/WAI/WCAG21/Understanding/labels-or-instructions.html

### PR DESCRIPTION
**Pull request by foxbunny**
- Introduce id and for attributes on inputs and labels
- Introduce aria-describedby for AccompanyingPeronInput button to provide additional context for the button
- Add a renderAsFieldset meta for registered fields which causes the fields to be renedered inside a fieldset with a legend
- Suppress the aria-label for react-dates datepicker as it interferes with the label associated with the control

Fixes #6059, fixes #6073

### CODE REVIEW

**Comment by foxbunny** [comment]
> guess making the label for a dropdown (e.g. for the user's title) clickable is not possible and thus not done on purpose?

Yes, it's a limitation of the Semantic UI dropdown implementation.

**Comment by foxbunny** [comment]
> in any case, can you fix the linter warnings and rebase please?

Are you referring to the default value linter warning?

**Comment by ThiefMaster** [comment]
Yes. Basically set the ID to something (can be null or undefined) in the defaultProps if you don't always pass a value, or make it required if you do

**Comment by foxbunny** [comment]
> Yes. Basically set the ID to something (can be null or undefined) in the defaultProps if you don't always pass a value, or make it required if you do

Ok. It kinda felt silly for me to say "undefined" if that's the default "default".

**Comment by ThiefMaster** [comment]
Yeah, I think it's one of those linter rules I might want to revisit some day...

**Comment by foxbunny** [comment]
Maybe we should just ignore the warning. I fear that adding these unnecessary defaults will just create unnecessary noise in the code, you know.

**Comment by foxbunny** [comment]
If I need to choose between linter noise and code noise, I'd prefer the linter noise, that is.

**Comment by ThiefMaster** [comment]
The problem is that it adds noise in every single PR... So I really prefer not to have unmuted linter warnings. ;x

FWIW I would absolutely not mind to disable just that part of the rule, but I'd also lose the "required props may *not* have default" which is covered by the [same rule](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/require-default-props.md) - and I think people setting a default for a required prop makes no sense, so that one should cause a linter warning.

**Comment by foxbunny** [comment]
So these liter rules are run against the *entire* code base every time you do a PR?

**Comment by ThiefMaster** [comment]
Yes, and at some point GitHub started showing annotations for code that's not touched in he PR as well...

BTW, I just opened jsx-eslint/eslint-plugin-react#3660 to suggest making the rule configurable in a way to get rid of the noise.

**Comment by foxbunny** [comment]
Also fixed two unrelated warnings from before this PR.

**Comment by ThiefMaster** [comment]
BTW unless I'm missing something you *always* pass an `htmlId`:

```jsx
  const htmlId = `input-${inputProps.id}`;
```

```jsx
      <InputComponent
        isRequired={inputRequired}
        disabled={disabled}
        isPurged={showPurged}
        htmlId={htmlId}
        {...inputProps}
      />
```

So if that's indeed the case, then IMHO the prop should be marked as `.isRequired` in the propTypes and then there's no need for a default anymore.

**Comment by foxbunny** [comment]
Done

File: indico/modules/events/registration/client/js/form/fields/AccommodationInput.jsx [hunk 1]
```diff
@@ -175,6 +177,7 @@ function AccommodationInputComponent({
 }
 
 AccommodationInputComponent.propTypes = {
+  id: PropTypes.string,
```
Comment (ThiefMaster) [inline indico/modules/events/registration/client/js/form/fields/AccommodationInput.jsx:180]: This one can probably be made required as well

File: indico/modules/events/registration/client/js/form/fields/AccompanyingPersonsInput.jsx [hunk 2]
```diff
@@ -183,7 +185,7 @@ function AccompanyingPersonsComponent({
           </Label>
         )}
         {placesLimit !== Infinity && (
-          <div styleName="places-left">
+          <div id={id ? `${id}-placeslimit` : ''} styleName="places-left">
```
Comment (ThiefMaster) [inline indico/modules/events/registration/client/js/form/fields/AccompanyingPersonsInput.jsx:188]: `id=""` feels wrong. maybe `undefined` as a fallback is better? (but again, if we always have an ID you don't need the ternary there at all)

File: indico/modules/events/registration/client/js/form/FormItem.jsx [hunk 3]
```diff
@@ -130,14 +145,24 @@ export default function FormItem({
...
             <Form.Field required={showAsRequired} styleName="field">
-              <label style={{opacity: disabled ? 0.8 : 1, display: 'inline-block'}}>{title}</label>
-              {retentionPeriodIcon}
-              <InputComponent
-                isRequired={inputRequired}
-                disabled={disabled}
-                isPurged={showPurged}
-                {...inputProps}
-              />
+              {meta.renderAsFiedset ? (
```
[the remark this answers is not in the document]
Comment (foxbunny) [reply indico/modules/events/registration/client/js/form/FormItem.jsx:148]: Hmmm

[the remark this answers is not in the document]
Comment (foxbunny) [reply indico/modules/events/registration/client/js/form/FormItem.jsx]: Fixed

File: indico/modules/events/registration/client/js/form/fields/DateInput.jsx [hunk 4]
```diff
@@ -88,10 +91,18 @@ DateInputComponent.propTypes = {
 };
 
 DateInputComponent.defaultProps = {
+  id: undefined,
```
Comment (ThiefMaster) [inline indico/modules/events/registration/client/js/form/fields/DateInput.jsx:94]: Intentional or missed when changing the ids to required?

File: indico/modules/events/registration/client/js/form/fields/MultiChoiceInput.jsx [hunk 5]
```diff
@@ -178,8 +181,13 @@ MultiChoiceInputComponent.propTypes = {
   existingValue: PropTypes.objectOf(PropTypes.number).isRequired,
 };
 
+MultiChoiceInputComponent.defaultProps = {
+  id: undefined,
+};
```
Comment (ThiefMaster) [inline indico/modules/events/registration/client/js/form/fields/MultiChoiceInput.jsx:186]: likewise as in my comment above

File: indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx [hunk 6]
```diff
@@ -148,7 +152,12 @@ SingleChoiceDropdown.propTypes = {
   existingValue: PropTypes.objectOf(PropTypes.number).isRequired,
 };
 
+SingleChoiceDropdown.defaultProps = {
+  id: undefined,
+};
```
Comment (ThiefMaster) [inline indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:157]: same here

File: indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx [hunk 7]
```diff
@@ -345,8 +365,13 @@ SingleChoiceInputComponent.propTypes = {
   existingValue: PropTypes.objectOf(PropTypes.number).isRequired,
 };
 
+SingleChoiceInputComponent.defualtProps = {
+  id: undefined,
+};
```
Comment (ThiefMaster) [inline indico/modules/events/registration/client/js/form/fields/SingleChoiceInput.jsx:370]: and here

File: indico/modules/events/registration/client/js/form/fields/DateInput.jsx (hunk 4 above)
Comment (foxbunny) [reply indico/modules/events/registration/client/js/form/fields/DateInput.jsx:94]: Well, do you want to require those ids? If so, then we can get rid of them.

Comment (ThiefMaster) [reply indico/modules/events/registration/client/js/form/fields/DateInput.jsx:94]: I mean if an id prop is always passed,I think it makes sense to mark them as required.

At least I assume that a prop which is not required isn't always provided...

**Pull request merged**